File: tests/unit/server/shell-integration.test.ts
import { mkdtempSync, readFileSync } from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import { beforeAll, describe, expect, it } from 'vitest';
import { shellIntegration, writeIntegrationFiles } from '../../../server/src/local/shell-integration.js';

let root: string;

beforeAll(() => {
  root = mkdtempSync(path.join(os.tmpdir(), 'muxus-shellint-'));
  writeIntegrationFiles(root);
});

describe('writeIntegrationFiles', () => {
  it('emits OSC 133 command reports from the zsh shim', () => {
    const zshrc = readFileSync(path.join(root, 'zsh/.zshrc'), 'utf8');
    expect(zshrc).toContain(String.raw`\e]133;D;%s\a`);
    expect(zshrc).toContain(String.raw`\e]133;C\a`);
    expect(zshrc).toContain(String.raw`\e]133;P;Cwd=%s\a`);
    expect(zshrc).toContain('__muxus_report_cwd');
    expect(zshrc).toContain('add-zsh-hook precmd');
    expect(zshrc).toContain('add-zsh-hook preexec');
  });

  it('sources the user zsh startup files before hooking', () => {
    const zshenv = readFileSync(path.join(root, 'zsh/.zshenv'), 'utf8');
    const zshrc = readFileSync(path.join(root, 'zsh/.zshrc'), 'utf8');
    expect(zshenv).toContain('${ZDOTDIR:-$HOME}/.zshenv');
    expect(zshrc).toContain('${ZDOTDIR:-$HOME}/.zshrc');
  });

  it('reports via PS0 and PROMPT_COMMAND in the bash shim', () => {
    const bash = readFileSync(path.join(root, 'bash-init.bash'), 'utf8');
    expect(bash).toContain('"$HOME/.bashrc"');
    expect(bash).toContain('PS0="\\[\\e]133;C\\a\\]${PS0-}"');
    expect(bash).toContain('PROMPT_COMMAND="__muxus_prompt_mark${PROMPT_COMMAND:+;$PROMPT_COMMAND}"');
    expect(bash).toContain(String.raw`\e]133;P;Cwd=%s\a`);
  });
});

describe('shellIntegration', () => {
  it('redirects zsh through the shim ZDOTDIR', () => {
    const result = shellIntegration('/usr/bin/zsh', {}, root);
    expect(result.args).toEqual([]);
    expect(result.env).toEqual({ ZDOTDIR: path.join(root, 'zsh') });
  });

  it('carries the user ZDOTDIR through for the shim to restore', () => {
    const result = shellIntegration('/bin/zsh', { ZDOTDIR: '/home/u/.config/zsh' }, root);
    expect(result.env).toEqual({
      ZDOTDIR: path.join(root, 'zsh'),
      MUXUS_USER_ZDOTDIR: '/home/u/.config/zsh',
    });
  });

  it('starts bash with the init-file shim', () => {
    const result = shellIntegration('/bin/bash', {}, root);
    expect(result.args).toEqual(['--init-file', path.join(root, 'bash-init.bash')]);
    expect(result.env).toEqual({});
  });

  it('prefixes the native cmd prompt with a live Windows cwd report', () => {
    const result = shellIntegration(
      String.raw`C:\Windows\System32\cmd.exe`,
      { PROMPT: '$S$P$G' },
      null,
      'win32',
    );

    expect(result.args).toEqual([]);
    expect(result.env).toEqual({ PROMPT: String.raw`$E]133;P;Cwd=$P$E\$S$P$G` });
  });

  it('uses the standard visible cmd prompt when none was configured', () => {
    expect(shellIntegration('cmd.exe', {}, null, 'win32').env).toEqual({
      PROMPT: String.raw`$E]133;P;Cwd=$P$E\$P$G`,
    });
  });

  it('leaves other Windows shells untouched', () => {
    expect(shellIntegration('powershell.exe', {}, null, 'win32')).toEqual({ args: [], env: {} });
  });

  it('leaves other shells untouched', () => {
    expect(shellIntegration('/usr/bin/fish', {}, root)).toEqual({ args: [], env: {} });
  });

  it('degrades to no integration when the shim dir is unavailable', () => {
    expect(shellIntegration('/bin/zsh', {}, null)).toEqual({ args: [], env: {} });
  });
});

File: app/internal/localpty/shellintegration_test.go
package localpty

import (
	"os"
	"path/filepath"
	"reflect"
	"strings"
	"testing"
)

// Cases ported from tests/unit/server/shell-integration.test.ts.

func writtenRoot(t *testing.T) string {
	t.Helper()
	root := t.TempDir()
	if err := WriteIntegrationFiles(root); err != nil {
		t.Fatal(err)
	}
	return root
}

func readShim(t *testing.T, root, rel string) string {
	t.Helper()
	data, err := os.ReadFile(filepath.Join(root, rel))
	if err != nil {
		t.Fatal(err)
	}
	return string(data)
}

func TestZshShimEmitsOSC133CommandReports(t *testing.T) {
	root := writtenRoot(t)
	zshrc := readShim(t, root, "zsh/.zshrc")
	for _, want := range []string{`\e]133;D;%s\a`, `\e]133;C\a`, "add-zsh-hook precmd", "add-zsh-hook preexec"} {
		if !strings.Contains(zshrc, want) {
			t.Fatalf(".zshrc missing %q", want)
		}
	}
}

func TestZshShimSourcesUserStartupFilesFirst(t *testing.T) {
	root := writtenRoot(t)
	if !strings.Contains(readShim(t, root, "zsh/.zshenv"), "${ZDOTDIR:-$HOME}/.zshenv") {
		t.Fatal(".zshenv must source the user .zshenv")
	}
	if !strings.Contains(readShim(t, root, "zsh/.zshrc"), "${ZDOTDIR:-$HOME}/.zshrc") {
		t.Fatal(".zshrc must source the user .zshrc")
	}
}

func TestBashShimReportsViaPS0AndPromptCommand(t *testing.T) {
	root := writtenRoot(t)
	bash := readShim(t, root, "bash-init.bash")
	for _, want := range []string{
		`"$HOME/.bashrc"`,
		`PS0="\[\e]133;C\a\]${PS0-}"`,
		`PROMPT_COMMAND="__muxus_prompt_mark${PROMPT_COMMAND:+;$PROMPT_COMMAND}"`,
	} {
		if !strings.Contains(bash, want) {
			t.Fatalf("bash-init.bash missing %q", want)
		}
	}
}

func TestZshRedirectedThroughShimZdotdir(t *testing.T) {
	root := writtenRoot(t)
	result := Integration("/usr/bin/zsh", map[string]string{}, root)
	if len(result.Args) != 0 {
		t.Fatalf("args = %v, want empty", result.Args)
	}
	want := map[string]string{"ZDOTDIR": filepath.Join(root, "zsh")}
	if !reflect.DeepEqual(result.Env, want) {
		t.Fatalf("env = %v, want %v", result.Env, want)
	}
}

func TestUserZdotdirCarriedThrough(t *testing.T) {
	root := writtenRoot(t)
	result := Integration("/bin/zsh", map[string]string{"ZDOTDIR": "/home/u/.config/zsh"}, root)
	want := map[string]string{
		"ZDOTDIR":            filepath.Join(root, "zsh"),
		"MUXUS_USER_ZDOTDIR": "/home/u/.config/zsh",
	}
	if !reflect.DeepEqual(result.Env, want) {
		t.Fatalf("env = %v, want %v", result.Env, want)
	}
}

func TestBashStartsWithInitFileShim(t *testing.T) {
	root := writtenRoot(t)
	result := Integration("/bin/bash", map[string]string{}, root)
	if !reflect.DeepEqual(result.Args, []string{"--init-file", filepath.Join(root, "bash-init.bash")}) {
		t.Fatalf("args = %v", result.Args)
	}
	if len(result.Env) != 0 {
		t.Fatalf("env = %v, want empty", result.Env)
	}
}

func TestOtherShellsUntouched(t *testing.T) {
	root := writtenRoot(t)
	result := Integration("/usr/bin/fish", map[string]string{}, root)
	if len(result.Args) != 0 || len(result.Env) != 0 {
		t.Fatalf("fish must spawn untouched, got %+v", result)
	}
}

func TestDegradesWithoutShimDir(t *testing.T) {
	result := Integration("/bin/zsh", map[string]string{}, "")
	if len(result.Args) != 0 || len(result.Env) != 0 {
		t.Fatalf("missing shim dir must disable integration, got %+v", result)
	}
}
